lib/portlayer/exec/container_cache.go

```diff
@@ -79,3 +109,106 @@ func (conCache *containerCache) Remove(idOrRef string) {
+		if newState != container.State {
+			switch newState {
+			case StateStopping:
+				// set state only in cache -- allow tether / tools to update vmx
+				log.Debugf("Container(%s) Shutdown via OOB activity", container.ExecConfig.ID)
+				container.State = StateStopped
+			case StateRunning:
+				// set state only in cache -- allow tether / tools to update vmx
+				log.Debugf("Container(%s) started via OOB activity", container.ExecConfig.ID)
+				container.State = newState
+			case StateStopped:
+				log.Debugf("Container(%s) stopped via OOB activity", container.ExecConfig.ID)
+				container.State = newState
+				// a powered off event will by pass guest shutdown, so we need to update
+				// the exit status
+				containerUpdate(session, container, "", -1)
+			case StateSuspended:
+				//set state only in cache
+				log.Debugf("Container(%s) suspened via OOB activity", container.ExecConfig.ID)
+				container.State = newState
+			case StateRemoved:
+				log.Debugf("Container(%s) removed via OOB activity", container.ExecConfig.ID)
+				// remove from cache
+				containers.Remove(container.ExecConfig.ID)
+
+			}
+		}
+	}
+
+	return
+}
+
+func containerUpdate(session *session.Session, container *Container, start string, exitStatus int) {
+
+	ctx, _ := context.WithTimeout(context.TODO(), propertyCollectorTimeout)
+
+	// this was an OOB update, so we're going to update the last know state which is in the cache
+	// and persist that to the VMX -- that should avoid information being lost due to PowerOff, etc
+	id := container.ExecConfig.ID
+	// update the exec config
+	container.Lock()
+	cc := container.ExecConfig.Sessions[id]
+	cc.Started = start
+	cc.ExitStatus = exitStatus
+	container.ExecConfig.Sessions[id] = cc
+	container.Unlock()
+
+	// get a handle and commit
+	h := container.newHandle()
+	wait := int32(30)
+	h.Commit(ctx, session, &wait)
+
+}
+
+// eventedState will determine the target container
+// state based on the current container state and the event
+func eventedState(e string, current State) State {
+	switch e {
+	case event.ContainerShutdown:
+		// no need to worry about existing state
+		return StateStopping
+	case event.ContainerPoweredOn:
+		// are we in the process of starting
+		if current != StateStarting {
+			// TODO: this sets the state correctly, but doesn't
+			// actually wait for the container process to start - we could see issues w/ps
+			//
+			// if we start a propery collector w/o a timeout might hit issue seen where
+			// OOB activity results in a loop at startup w/o container process ever starting
+			return StateRunning
+		}
+	case event.ContainerPoweredOff:
+		// are we in the process of stopping
+		if current != StateStopping {
+			return StateStopped
+		}
+	case event.ContainerSuspended:
+		// are we in the process of suspending
+		if current != StateSuspending {
+			return StateSuspended
+		}
+	case event.ContainerRemoved:
+		if current != StateRemoving {
+			return StateRemoved
```

Comment:
Is this essentially asking after an event when the vm is no long in the transition state, assume it is in the final state?  Isn't that inherently quite risky?  If you attempt power on, it is in starting state, but then fails and enters poweredoff state, would we then report it is in the running state?